homeassistant/components/omnilogic/sensor.py

```diff
@@ -0,0 +1,554 @@
+                            bow["CSAD"],
+                        )
+                    )
+
+    async_add_entities(sensors, update_before_add=True)
+
+
+class OmnilogicSensor(Entity):
+    """Defines an Omnilogic sensor entity."""
+
+    def __init__(
+        self,
+        coordinator,
+        kind,
+        name,
+        backyard,
+        bow,
+        device_class,
+        icon,
+        unit,
+        sensordata,
+    ):
+        """Initialize Entities."""
+
+        if bow != {}:
+            sensorname = (
+                backyard["BackyardName"].replace(" ", "_")
+                + "_"
+                + bow["Name"].replace(" ", "_")
+                + "_"
+                + kind
+            )
+        else:
+            sensorname = backyard["BackyardName"].replace(" ", "_") + "_" + kind
+
+        self._kind = kind
+        self._name = None
+        self.entity_id = ENTITY_ID_FORMAT.format(sensorname)
+        self._backyard = backyard
+        self._backyard_name = backyard["BackyardName"]
+        self._state = None
+        self._unit_type = backyard["Unit-of-Measurement"]
+        self._device_class = device_class
+        self._icon = icon
+        self._bow = bow
+        self._unit = None
+        self.coordinator = coordinator
+        self.bow = bow
+        self.sensordata = sensordata
+        self.attrs = {}
+        self.attrs["MspSystemId"] = backyard["systemId"]
+        self.alarms = []
+
+    @property
+    def should_poll(self) -> bool:
+        """Return the polling requirement of the entity."""
+        return True
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique, Home Assistant friendly identifier for this entity."""
+        return self.entity_id
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def device_class(self):
+        """Return the device class of the entity."""
+        if self._device_class != "none":
+            return self._device_class
+
+    @property
+    def unit_of_measurement(self):
+        """Return the right unit of measure."""
+        return self._unit
+
+    @property
+    def icon(self):
+        """Return the icon for the entity."""
+        return self._icon
+
+    @property
+    def device_state_attributes(self):
+        """Return the attributes."""
+        return self.attrs
+
+    @property
+    def force_update(self):
+        """Force update."""
+        return True
+
+    @property
+    def state(self):
+        """Return the state."""
+        return self._state
+
+    @property
+    def device_info(self):
+        """Define the device as back yard/MSP System."""
+
+        return {
+            "identifiers": {(DOMAIN, self.attrs["MspSystemId"])},
+            "name": self._backyard.get("BackyardName"),
+            "manufacturer": "Hayward",
+            "model": "OmniLogic",
+        }
+
+    async def async_update(self):
```

Comment:
This should all be in the coordinator callback.

Reply:
There's also a lot of logic here, and it's really hard to follow. Does it make sense to break this into several sensor classes with one common base? That way the code for each class contains less logic.

Reply:
Addressed the coordinator callback issue. Will be breaking the sensors into unique classes to reduce logic/complexity tomorrow.